docs/design.md

```diff
@@ -0,0 +1,547 @@
+  HkvVwA==
+  -----END CERTIFICATE-----
+...
+```
+
+#### CertificateKey
+
+```yaml
+---
+schema: deckhand/CertificateKey/v1
+metadata:
+  schema: metadata/Document/v1
+  name: application-api
+  storagePolicy: encrypted
+data: |-
+  -----BEGIN RSA PRIVATE KEY-----                                 
+  MIIEpQIBAAKCAQEAx+m1+ao7uTVEs+I/Sie9YsXL0B9mOXFlzEdHX8P8x4nx78/T
+  ...snip...
+  Zf3ykIG8l71pIs4TGsPlnyeO6LzCWP5WRSh+BHnyXXjzx/uxMOpQ/6I=
+  -----END RSA PRIVATE KEY-----
+...
+```
+
+#### Passphrase
+
+```yaml
+---
+schema: deckhand/Passphrase/v1
+metadata:
+  schema: metadata/Document/v1
+  name: application-admin-password
+  storagePolicy: encrypted
+data: some-password
+...
+```
+
+## Revision History
+
+Documents will be ingested in batches which will be given a revision index.
+This provides a common language for describing complex validations on sets of
+documents.
+
+Revisions can be thought of as commits in a linear git history, thus looking
+at a revision includes all content from previous revision.
+
+## Validation
+
+Services can report success for validation types for a given revision.
+
+## API
+
+This API will only support YAML as a serialization format. Since the IETF
+does not provide an official media type for YAML, this API will use
+`application/x-yaml`.
+
+### POST `/documents`
+
+Accepts a multi-document YAML body and creates a new revision which adds
+those documents. Updates are detected based on exact match to an existing
+document of `schema` + `metadata.name`. Documents are "deleted" by including
+documents with the tombstone metadata schema, such as:
+
+```yaml
+---
+schema: any-namespace/AnyKind/v1
+metadata:
+  schema: metadata/Tombstone/v1
+  name: name-to-delete
+...
+```
+
+This endpoint is the only way to add, update, and delete documents. This
+triggers Deckhand's internal schema validations for all documents.
+
+If no changes are detected, a new revision should not be created. This allows
+services to periodically re-register their schemas without creating
+unnecessary revisions.
+
+### GET `/revisions/{revision_id}/documents`
+
+Returns a multi-document YAML response containing all the documents matching
+the filters specified via query string parameters. Returned documents will be
+fully merged and substituted.
+
+Supported query string parameters:
+
+* `schema` - string, optional - The top-level `schema` field to select. This
+  may be partially specified by section, e.g., `schema=promenade` would select all
+  `kind` and `version` schemas owned by promenade, or `schema=promenade/Node`
+  which would select all versions of `promenade/Node` documents. One may not
+  partially specify the namespace or kind, so `schema=promenade/No` would not
+  select `promenade/Node/v1` documents, and `schema=prom` would not select
+  `promenade` documents.
+* `metadata.name` - string, optional
+* `metadata.layeringDefinition.abstract` - string, optional - Valid values are
```

Comment:
since we are POSTing documents, it seems like GET documents should return those things being posted (Doesn't have to, but it is user friendly to do so)- so I agree, a new endpoint for compiledDocuments seems like a very reasonable thing to GET.  I also think it's reasonable that a compileDocument include references somehow to the list of all documents that went into that compilation.